tests/test_data/packages/small_fake_with_build_deps/backend/backend.py

```diff
@@ -0,0 +1,24 @@
+from __future__ import annotations
+
+import setuptools.build_meta
+
+build_wheel = setuptools.build_meta.build_wheel
+build_sdist = setuptools.build_meta.build_sdist
+
+
+def get_requires_for_build_sdist(config_settings=None):
+    result = setuptools.build_meta.get_requires_for_build_sdist(config_settings)
+    assert result == []
+    result.append("small-fake-a")
+    return result
+
+
+def get_requires_for_build_wheel(config_settings=None):
+    result = setuptools.build_meta.get_requires_for_build_wheel(config_settings)
+    assert result == ["wheel"]
+    result.append("small-fake-b")
```

Comment:
I was recently reminded that there's also a case of the hook-returned deps requiring those declared in `pyproject.toml`. It'd be interesting to see such a dependency (besides setuptools) imported at the top of the in-tree backend module.

Reply:
I am not sure I follow. Are you describing a situation where
* `build-system.requires` requires package a,
* `get_requires_for_build_wheel` requires package b, and
* package b has a runtime dependency on package a?

Reply:
@apljungquist sorry, I didn't reply sooner.

Imagine the following build backend module:

```python
# build_backend.py
...
from contextlib import suppress

with suppress(ImportError):
    import b
...

def get_requires_for_build_wheel(config_settings=None):
    wheel_build_requirements = ['b']
    with suppress(NameError):
        b
        wheel_build_requirements.append('c')
    return wheel_build_requirements
```
and the following declarative config:
```toml
# pyproject.toml
[build-system]
requires = ["a"]
build-backend = "build_backend"
```

Now, a build front-end would work in the following order:
1. Populate package `a`.
2. Invoke `get_requires_for_build_wheel()` which only returns `['b']` on that first iteration since at this point, `b` is not importable so the other dependency isn't returned.
3. Populate `b` into the build venv.
4. Invoke `get_requires_for_build_wheel()`, which now returns `['b', 'c']` because `b` was importable at this iteration and `NameError` didn't happen.
5. Invoke `get_requires_for_build_wheel()`, see no difference, populate nothing, stop calling the hook.
6. At this point, the collected build requirements would be `['a', 'b', 'c']`.

This is basically the case I wanted to make sure works — recursive augmentation of the build deps.

Reply:
I've just checked PEP 517, and it doesn't seem to specify if `get_requires_for_build_wheel()` can be called multiple times. So maybe, we could go for something simpler, like:

```python
# build_backend.py
...
from contextlib import suppress

with suppress(ImportError):
    import b
...

def get_requires_for_build_wheel(config_settings=None):
    return ['b']
```
and the following declarative config:
```toml
# pyproject.toml
[build-system]
requires = ["a"]
build-backend = "build_backend"
```

Reply:
> Aren't we testing the build front end at this point?

Well, we _do act_ as a build front end in this case, don't we?

Reply:
I don't think I understand what it is that you are requesting. We have a test case ensuring that `pyproject.toml::build-system.requires` are installed before calling `get_requires_for_build_wheel`. And since `get_requires_for_build_wheel` is called at most once it's output cannot not affect its output.

Reply:
@apljungquist

> And since `get_requires_for_build_wheel` is called at most once it's output cannot not affect its output.

This is where the confusion lies: PEP 517 doesn't _explicitly_ state that `get_requires_for_build_wheel` can only be called once. So I thought that maybe it's a good idea to cover the case where installing the returned deps from the first invocation and calling it again could return more requirements that depend on the first chunk of deps…

Where did I get such idea from? A similar thing is implemented in Fedora — https://fedoraproject.org/wiki/Changes/DynamicBuildRequires#Example. They essentially made it possible for packaging specs to generate build deps dynamically. But if there are unsatisfied deps, they restart generation again, and it can output more things on the new iterations.

I figured, since it seems to be a gray area right now, PEP-wise, then it's worth supporting, maybe…

Reply:
Okay, let's declare this out of the scope. Marking this thread as resolved.